spec/lib/auth/managed_authenticator_spec.rb

```diff
@@ -254,6 +254,66 @@ def match_by_email
         expect(user.user_profile.location).to eq("DiscourseVille")
       end
     end
+
+    describe 'match by username' do
+      let(:user_match_authenticator) {
+        Class.new(described_class) do
+          def name
+            "myauth"
+          end
+          def match_by_email
+            false
+          end
+          def match_by_username
+            true
+          end
+        end.new
+      }
+
+      it 'works normally' do
+        SiteSetting.username_change_period = 0
+        user = Fabricate(:user)
+        result = user_match_authenticator.after_authenticate(hash.deep_merge(info: { nickname: user.username }))
+        expect(result.user.id).to eq(user.id)
+        expect(UserAssociatedAccount.find_by(provider_name: 'myauth', provider_uid: "1234").user_id).to eq(user.id)
+      end
+
+      it 'works if there is already an association with the target account' do
+        SiteSetting.username_change_period = 0
+        user = Fabricate(:user, username: "IAmGroot")
+        result = user_match_authenticator.after_authenticate(hash)
+        expect(result.user.id).to eq(user.id)
+      end
+
+      it 'works if the username is different case' do
+        SiteSetting.username_change_period = 0
+        user = Fabricate(:user, username: "IAMGROOT")
+        result = user_match_authenticator.after_authenticate(hash)
+        expect(result.user.id).to eq(user.id)
+      end
+
+      it 'does not match if username_change_period isn\'t 0' do
+        SiteSetting.username_change_period = 3
+        user = Fabricate(:user, username: "IAmGroot")
+        result = user_match_authenticator.after_authenticate(hash)
+        expect(result.user).to eq(nil)
+      end
+
+      it 'does not match if match_by_username is false' do
```

Comment:
Can we add another test like "does not match if match_by_username is the default", and create a class which doesn't override `def match_by_username`?

(if the default was accidently changed, that would be very bad, so it would be good to have a test for it)

Reply:
Added